File: src/main/java/me/delivery/domain/address/model/entity/Address.java
package me.delivery.domain.address.model.entity;

import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import me.delivery.domain.address.model.vo.AddressVO;
import org.springframework.data.geo.Point;

import javax.persistence.*;
import java.math.BigDecimal;

@Entity(name = "address")
@Getter
@Setter
@NoArgsConstructor
public class Address {
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    private long id;
    private String address;
    private String metropolitan;
    private String city;
    private String town;
    private String road;
    private Integer mainNumber;
    private Integer subNumber;
    @Column(name = "point", columnDefinition = "POINT")
    private Point point;

    @Builder
    private Address(
        String address,
        String metropolitan,
        String city,
        String town,
        String road,
        Integer mainNumber,
        Integer subNumber,
        BigDecimal x,
        BigDecimal y
    ) {
        this.address = address;
        this.metropolitan = metropolitan;
        this.city = city;
        this.town = town;
        this.road = road;
        this.mainNumber = mainNumber;
        this.subNumber = subNumber;
        this.point = new Point(x.doubleValue(), y.doubleValue());
    }

    public AddressVO toImmutable() {
        return AddressVO.builder()
                .id(id)
                .address(address)
                .metropolitan(metropolitan)
                .city(city)
                .town(town)
                .road(road)
                .mainNumber(mainNumber)
                .subNumber(subNumber)
                .point(point)
                .build();
    }
}

File: src/main/kotlin/me/delivery/domain/address/model/entity/Address.kt
package me.delivery.domain.address.model.entity

import me.delivery.domain.address.model.vo.AddressVO
import me.delivery.domain.entity.BaseEntity
import org.springframework.data.geo.Point
import java.math.BigDecimal
import javax.persistence.*
import kotlin.jvm.Transient

@Entity(name = "address")
class Address (
    val address: String,
    val metropolitan: String,
    val city: String,
    val town: String,
    val road: String,
    val mainNumber: Int,
    val subNumber: Int,
    @Transient
    val x: BigDecimal,
    @Transient
    val y: BigDecimal,
): BaseEntity() {

    @Column(name = "point", columnDefinition = "POINT")
    val point: Point = Point(x.toDouble(), y.toDouble())

    fun toImmutable(): AddressVO = AddressVO(this)
}